Report the presence of formulae in Lotus Symphony-based spreadsheets

codeofdusk opened the pull request:
<!--
Please fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/wiki/Github-pull-request-template-explanation-and-examples
-->

### Link to issue number:
#860

### Summary of the issue:
Unlike in Excel, NVDA does not announce the presence of formulae in Lotus Symphony spreadsheets, such as in [LibreOffice](http://libreoffice.org) or [OpenOffice](http://openoffice.org).

### Description of how this pull request fixes the issue:
As documented in the issue, these applications now expose an IAccessible 2 attribute, Formula, which contains the formula of a given cell if present. I've added a check in `appModules.soffice.SymphonyTableCell._get_states` for this attribute: if it indicates a formula for the cell, its `ControlTypes.STATE_HASFORMULA` is set.

### Testing performed:
Read the [attached spreadsheet](https://github.com/nvaccess/nvda/files/2719061/plants.xlsx) using LibreOffice Calc on Windows 10 1809 before and after the patch. Noted that "has formula" is announced after the patch for cells containing formulae.

### Known issues with pull request:
None.

### Change log entry:

== New Features ==
- The presence of formulae in LibreOffice and Apache OpenOffice spreadsheets is now reported. (#860)

codeofdusk committed:
Fixed #860

This commit fixes #860 by checking the "Formula" IAccessible2 attribute exposed by Lotus Symphany based spreadsheet applications and adding the STATE_HASFORMULA where needed.

josephsl commented:
Hi,

Thanks for this one.

Can you guarantee that this will work across LibreOffice versions (not just dev, fresh, and stil)?

Thanks.

josephsl reviewed (commented):
Hi,

Is tones module used at all in appModules/soffice.py?

Also, I'd like to suggest adding an explanatory comment about adding "has formula" state. Some like:

"""# #860: recent LibreOffice versions exposes has formula state via IA2."""

Don't worry about the quotes (I did it to get around Markdown markup here).

I'd like to suggest creating a try build for LibreOffice users so they can test your changes in Calc in the new year.

Thanks.

codeofdusk commented:
Not sure – I tested on latest LibreOffice stable, and the attribute was mentioned in 2017, so must’ve existed since at least then. If it isn’t available, it shouldn’t cause any problems – merely formulae won’t be reported (NVDA’s behavior before the patch).

Bill 
Sent from my iPhone

> On Dec 31, 2018, at 20:51, Joseph Lee <notifications@github.com> wrote:
> 
> Hi,
> 
> Thanks for this one.
> 
> Can you guarantee that this will work across LibreOffice versions (not just dev, fresh, and stil)?
> 
> Thanks.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

josephsl commented:
Hi, ah, in this case, likely a 6.0.x release or a late 5.x release then. Don’t have the GitHub ID for our libreoffice expert at the moment… Thanks.

 

From: Bill Dengler <notifications@github.com> 
Sent: Monday, December 31, 2018 6:00 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Comment <comment@noreply.github.com>
Subject: Re: [nvaccess/nvda] Report the presence of formulae in Lotus Symphony-based spreadsheets (#9115)

 

Not sure – I tested on latest LibreOffice stable, and the attribute was mentioned in 2017, so must’ve existed since at least then. If it isn’t available, it shouldn’t cause any problems – merely formulae won’t be reported (NVDA’s behavior before the patch).

Bill 
Sent from my iPhone

> On Dec 31, 2018, at 20:51, Joseph Lee <notifications@github.com> wrote:
> 
> Hi,
> 
> Thanks for this one.
> 
> Can you guarantee that this will work across LibreOffice versions (not just dev, fresh, and stil)?
> 
> Thanks.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.



—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/9115#issuecomment-450704695> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkI009FAqZc97h4Hne5LwIZHi_DdNks5u-sEggaJpZM4Zluhd> .

codeofdusk commented:
@josephsl Oops, I imported the tones module while testing something and forgot to delete that line before committing. I've also added an explanatory comment as suggested and fixed up the commit. Additionally, I've tested the patch with Apache OpenOffice and confirmed that it works.

josephsl reviewed (approved):
Excellent, thanks. I'll just add proper copyright header, and that'll be it.

josephsl committed:
App modules/Lotus Symphoniy suite: updated copyright headers

leonardder commented on source/appModules/soffice.py:
```diff
@@ -220,6 +220,9 @@ def _get_states(self):
 			# #8988: Cells in Libre Office do not have the selected state when a single cell is selected (i.e. has focus).
 			# Since #8898, the negative selected state is announced for table cells with the selectable state.
 			states.add(controlTypes.STATE_SELECTED)
+		if 'Formula' in self.IA2Attributes and self.IA2Attributes['Formula']:
```
```suggestion
 		if self.IA2Attributes.get('Formula'):
```

LeonarddeR reviewed (changes requested):
I think that it might be safe to consider applying this patch to NVDAObjects.IAccessible.IAccessible.

@MICHAELDCURRAN AND @DEREKRIEMER: Might it be an idea to discuss with Google whether they could expose a formula like this in Google Sheets?

leonardder commented:
> I think that it might be safe to consider applying this patch to NVDAObjects.IAccessible.IAccessible.

Hmm never mind, starting the word Formula with a capital F doesn't look like they're following some kind of a standard there.

LeonarddeR committed:
Update source/appModules/soffice.py

Co-Authored-By: codeofdusk <codeofdusk@gmail.com>

codeofdusk commented:
Oops, of course. Thanks @leonardder.

LeonarddeR reviewed (approved):
Great

michaelDCurran committed:
Update what's new.

josephsl commented:
Hi,

Hmmm, is there a specific reason why the word "formulae" was chosen as a plural form of "formula"? This may cause potential confusion for people reading the what's new. I normally do not edit what's new entries just to change plural forms, but I think a small clarification here might be useful.

Thanks.

codeofdusk commented:
No, no particular reason – [Wiktionary](
https://en.wiktionary.org/wiki/formula#Noun) says "formulae" and "formulas" are both acceptable as the plural form. Personally, in informal speech I might say "formulas" and in writing "formulae" but either form is acceptable to me.
